### PR TITLE
Add FTS algorithm

### PR DESCRIPTION
**Pull request by TheSunWillRise**
Add implementation of FTS algorithm [https://arxiv.org/abs/2010.10154](url)

### CODE REVIEW

File: federatedscope/autotune/algos.py [hunk 1]
```diff
@@ -398,6 +401,27 @@ def _setup(self):
...
+class FTSAlgo(Scheduler):
+    """FTS algorithm runs like a standard FL,
+        e.g. The server maintain a GP model for a client. At each
+        communication round, the server samples a config from the GP model
+         and broadcasts to the target client. A client training a moddel
+          from scratch on its local data with received hyper-param and return
+          the validation results to the server.
+    """
+
+    def _setup(self):
```
Comment (rayrayraykk) [inline federatedscope/autotune/algos.py:413]: We'd prefer to use `TrialExecutor` in https://github.com/alibaba/FederatedScope/blob/71bc828b4b6ad5d116de12c980b352500af314ac/federatedscope/autotune/algos.py#L38.

Besides, can we use FTS as a wrapper to wrap optimizers like FedEx here? You can refer to: https://github.com/alibaba/FederatedScope/blob/71bc828b4b6ad5d116de12c980b352500af314ac/federatedscope/autotune/algos.py#L322.

**Review by rayrayraykk — COMMENTED** [review]
Good job, please see the inline comments and make some modifications to execute FTS as a wrapper. With a wrap function, we could conduct RS+FTS, SHA+FTS, etc. You can refer to PR #311 for details.

For running wrapper FedEx:
`python federatedscope/hpo.py --cfg scripts/example_configs/femnist/avg/bo_kde_wrap.yaml`

File: federatedscope/autotune/algos.py (hunk 1 above)
Comment (joneswong) [reply federatedscope/autotune/algos.py:413]: I think FTS has a similar role as FedEx, what we want is to apply common HPO toolkit / our SHA to wrap FTS.

**Comment by TheSunWillRise** [comment]
I have add codes to support to apply HPO methods (e.g. SHA) to wrap FTS.
Now we can use sha to wrap fts by running 
`python federatedscope/hpo.py --cfg scripts/example_configs/femnist/sha_wrap_fts.yaml`.
If we want to run fts for only searching clients' hyper-parameters as the experiments in the paper, we can run
'python federatedscope/main.py --cfg scripts/example_configs/femnist/avg/fts.yaml'.

File: federatedscope/autotune/algos.py [hunk 2]
```diff
@@ -78,6 +79,8 @@ def get_scheduler(init_cfg):
     #     scheduler = PBT(init_cfg)
     elif init_cfg.hpo.scheduler.startswith('wrap'):
         scheduler = SHAWrapFedex(init_cfg)
+    elif init_cfg.hpo.scheduler == 'fts':
+        scheduler = FTSAlgo(init_cfg)
```
Comment (rayrayraykk) [inline federatedscope/autotune/algos.py:83]: No definition of `FTSAlgo`.

Comment (TheSunWillRise) [reply federatedscope/autotune/algos.py]: I just deleted everything about FTSAlgo. We don't need this algorithm now. All process of fts are contained in FTSServer and FTSClient

**Comment by TheSunWillRise** [comment]
I run a experiment on femnist to compare fts vs local bo.
The search space is lr: [0.005, o0.5], weight decay: [0.0, 1.0], dropout: [0.0, 0.5].
There are 5 total clients and  I search hyper-params for 3 of them.
The test accuracy of 3 clients are:
- FTS: 0.72973, 0.583333, 0.741935
- Local BO: 0.72973, 0.611111, 0.677419

Detailed results:

- FTS
- 
2022-09-05 11:08:19,852 (server:390)INFO: **************************************************
2022-09-05 11:08:19,852 (server:391)INFO: {'Role': 'Server #', 'Round': 'Final', 'Results_raw': {
1: {'test_total': [37], 'test_correct': [27.0], 'test_f1': [0.412434], 'test_acc': [0.72973], 'test_avg_loss': [2.134613], 'test_loss': [78.980685], 'val_total': [36], 'val_correct': [28.0], 'val_f1': [0.432353], 'val_acc': [0.777778], 'val_avg_loss': [1.808752], 'val_loss': [65.115088], 'train_total': [109], 'train_correct': [101.0], 'train_f1': [0.807304], 'train_acc': [0.926606], 'train_avg_loss': [0.65794], 'train_loss': [71.715504]}, 
2: {'test_total': [36], 'test_correct': [21.0], 'test_f1': [0.293375], 'test_acc': [0.583333], 'test_avg_loss': [2.911377], 'test_loss': [104.809578], 'val_total': [35], 'val_correct': [27.0], 'val_f1': [0.503128],'val_acc': [0.771429], 'val_avg_loss': [1.994824], 'val_loss': [69.818842], 'train_total': [105], 'train_correct': [78.0], 'train_f1': [0.321938], 'train_acc': [0.742857], 'train_avg_loss': [1.35125], 'train_loss': [141.881243]}, 
4: {'test_total': [31], 'test_correct': [23.0], 'test_f1': [0.45042], 'test_acc': [0.741935], 'test_avg_loss': [1.914418], 'test_loss': [59.346959], 'val_total': [30], 'val_correct': [15.0], 'val_f1': [0.237554], 'val_acc': [0.5], 'val_avg_loss': [3.244016], 'val_loss': [97.320467], 'train_total': [90], 'train_correct': [61.0], 'train_f1': [0.207175], 'train_acc': [0.677778], 'train_avg_loss': [1.400578], 'train_loss': [126.052037]}}}
2022-09-05 11:08:19,853 (server:397)INFO: **************************************************

- Local BO

2022-09-05 13:56:09,893 (server:390)INFO: **************************************************
2022-09-05 13:56:09,893 (server:391)INFO: {'Role': 'Server #', 'Round': 'Final', 'Results_raw': {
1: {'test_correct': [27.0], 'test_total': [37], 'test_f1': [0.364286], 'test_acc': [0.72973], 'test_avg_loss': [3.001294], 'test_loss': [111.047867], 'val_correct': [28.0], 'val_total': [36], 'val_f1': [0.440741], 'val_acc': [0.777778], 'val_avg_loss': [2.070544], 'val_loss': [74.539566], 'train_correct': [108.0], 'train_total': [109], 'train_f1': [0.969697], 'train_acc': [0.990826], 'train_avg_loss': [0.053973], 'train_loss': [5.883042]}, 
2: {'test_correct': [22.0], 'test_total': [36], 'test_f1': [0.27998], 'test_acc': [0.611111], 'test_avg_loss': [2.611234], 'test_loss': [94.00444], 'val_correct': [28.0], 'val_total': [35], 'val_f1': [0.490936], 'val_acc': [0.8], 'val_avg_loss': [1.666421], 'val_loss': [58.324727], 'train_correct': [96.0], 'train_total': [105], 'train_f1': [0.773136], 'train_acc': [0.914286], 'train_avg_loss': [0.59692], 'train_loss': [62.67657]}, 
4: {'test_correct': [21.0], 'test_total': [31], 'test_f1': [0.36468], 'test_acc': [0.677419], 'test_avg_loss': [2.163686], 'test_loss': [67.074264], 'val_correct': [11.0], 'val_total': [30], 'val_f1': [0.183858], 'val_acc': [0.366667], 'val_avg_loss': [3.320269], 'val_loss': [99.608057], 'train_correct': [59.0], 'train_total': [90], 'train_f1': [0.182847], 'train_acc': [0.655556], 'train_avg_loss': [1.618616], 'train_loss': [145.675462]}}}
2022-09-05 13:56:09,894 (server:397)INFO: **************************************************

File: federatedscope/autotune/algos.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from federatedscope.core.fed_runner import FedRunner
 from federatedscope.autotune.utils import parse_search_space, \
     config2cmdargs, config2str, summarize_hpo_results
+from federatedscope.autotune.fts import FTSClient, FTSServer
```
Comment (joneswong) [inline federatedscope/autotune/algos.py:19]: remove this line

File: federatedscope/core/configs/cfg_hpo.py [hunk 4]
```diff
@@ -81,5 +99,9 @@ def assert_hpo_cfg(cfg):
                                                        "federate.use_diff " \
                                                        "must be True."
 
+    assert cfg.hpo.fts.use == cfg.federate.use_diff, "Once FTS is " \
```
Comment (joneswong) [inline federatedscope/core/configs/cfg_hpo.py:102]: Suppose I want to conduct HPO by FedEx with the difference of validation performances before and after local training as the feedback, then `fits.use` is false and `use_diff` is true. This assertion seems to be invoked.

File: federatedscope/autotune/fts/utils.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
+from datetime import datetime
+import GPy
```
Comment (joneswong) [inline federatedscope/autotune/fts/utils.py:2]: The introduced dependency should be mentioned in the "requirement" declaration (see /environment/).

File: federatedscope/autotune/fts/utils.py [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+import pickle
+import time
+import numpy as np
+import warnings
+
+from scipydirect import direct
+
+warnings.filterwarnings('ignore')
+
+
```
Comment (joneswong) [inline federatedscope/autotune/fts/utils.py:12]: Code in this file seems to be largely borrowed from FTS's repo, right? If it is the case, please provide declaration at the beginning of this file and their copyright in /LICENSE.

File: scripts/example_configs/femnist/avg/fts.yaml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+criterion:
+  type: CrossEntropyLoss
+trainer:
+  type: cvtrainer
+eval:
+  freq: 1
+  metrics: ['acc', 'correct', 'f1']
+  split: ['test', 'val', 'train']
+hpo:
+  scheduler: 'fts'
```
Comment (joneswong) [inline scripts/example_configs/femnist/avg/fts.yaml:44]: Sorry for my mis-understanding of this work at that time. For now, is it necessary to have a FTS scheduler as the entry? If it is the case, where can I find the implementation (it seems that nothing in algo.py)?

**Review by joneswong — CHANGES_REQUESTED** [review]
Good job. it is great to see that code snippets from other repo can be easily integrated into FS. Please resolve the issues I point out. Thanks!
